starfish/util/overlap_utils.py

```diff
@@ -0,0 +1,162 @@
+from starfish.types._constants import OverlapStrategy
+
+
+class Area:
+    """
+    Small class that defines rectangular area of physical space by
+    its bottom left and top right coordinates.
+    """
+    def __init__(self, min_x, max_x, min_y, max_y):
+        self.min_x = min_x
+        self.max_x = max_x
+        self.min_y = min_y
+        self.max_y = max_y
+
+    def __eq__(self, other):
+        return (self.min_x == other.min_x
+                and self.min_y == other.min_y
+                and self.max_x == other.max_x
+                and self.max_y == other.max_y)
+
+    @staticmethod
+    def _overlap(area1: "Area", area2: "Area") -> bool:
+        """Return True if two rectangles overlap"""
+        if (area1.max_x < area2.min_x) or (area1.min_x > area2.max_x):
+            return False
+        if (area1.max_y < area2.min_y) or (area1.min_y > area2.max_y):
+            return False
+        return True
+
+    @staticmethod
+    def find_intersection(area1: "Area", area2: "Area") -> Union[None, "Area"]:
+        """
+        Find the overlap area of two rectangles and return as new Area object.
+        If no overlap return none.
+        """
+        if Area._overlap(area1, area2):
+            return Area(max(area1.min_x, area2.min_x),
+                        min(area1.max_x, area2.max_x),
+                        max(area1.min_y, area2.min_y),
+                        min(area1.max_y, area2.max_y))
+        return None
+
+
+def find_overlaps_of_xarrays(xarrays: List[xr.DataArray]
+                             ) -> Dict[Tuple[int, int], "Area"]:
+    """
+    Find all the overlap areas within a list of xarrays.
+
+    Parameters
+    ----------
+    xarrays : List[xr.DataArray]
+        The list of xarrays to find overlaps in.
+
+    Returns
+    -------
+    Dict[Tuple[int, int], "Area"]] :
+        A dictionary of tuples containing the indices of two overlapping
+        IntensityTables and their Area of intersection.
+
+    """
+    all_overlaps: Dict[Tuple[int, int], "Area"] = dict()
+    for idx1, idx2 in itertools.combinations(range(len(xarrays)), 2):
+        xr1 = xarrays[idx1]
+        xr2 = xarrays[idx2]
+        area1 = Area(min(xr1[Coordinates.X.value]).data,
+                     max(xr1[Coordinates.X.value]).data,
+                     min(xr1[Coordinates.Y.value]).data,
+                     max(xr1[Coordinates.Y.value]).data)
+
+        area2 = Area(min(xr2[Coordinates.X.value]).data,
+                     max(xr2[Coordinates.X.value]).data,
+                     min(xr2[Coordinates.Y.value]).data,
+                     max(xr2[Coordinates.Y.value]).data)
+        intersection = Area.find_intersection(area1, area2)
+        if intersection:
+            all_overlaps[(idx1, idx2)] = intersection
+    return all_overlaps
+
+
+def remove_area_of_xarray(it: xr.DataArray, area: Area) -> xr.DataArray:
+    """Return everything in the xarray defined OUTSIDE the input area
```

Comment:
It would be good to explain the boundary condition.  It looks like spots on the line are removed.